### PR TITLE
Add prune controller to the APIServerControllerSet

### PR DESCRIPTION
**Pull request by stlaz**
\+ a small bugfix

### CODE REVIEW

File: pkg/operator/apiserver/controllerset/apiservercontrollerset.go [hunk 1]
```diff
@@ -64,12 +65,13 @@ type APIServerControllerSet struct {
 	apiServiceController            controllerWrapper
 	clusterOperatorStatusController controllerWrapper
 	configUpgradableController      controllerWrapper
-	logLevelController              controllerWrapper
```
Comment (stlaz) [inline pkg/operator/apiserver/controllerset/apiservercontrollerset.go:67]: this is purely cosmetical change, sorting these alphabetically

File: pkg/operator/apiserver/controller/prune/prune_controller.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+	podInformer    corev1informer.PodInformer
+	secretInformer corev1informer.SecretInformer
+}
+
+const (
+	numOldRevisionsToPreserve = 5
+)
+
+// NewPruneController creates a new pruning controller
+func NewPruneController(
```
Comment (p0lyn0mial) [inline pkg/operator/apiserver/controller/prune/prune_controller.go:41]: NewSecretPruneController ?

File: pkg/operator/apiserver/controller/prune/prune_controller.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
+package prune
```
Comment (p0lyn0mial) [inline pkg/operator/apiserver/controller/prune/prune_controller.go:1]: is this controller specific to `apiserver` package?

File: pkg/operator/apiserver/controllerset/apiservercontrollerset.go [hunk 4]
```diff
@@ -275,7 +277,31 @@ func (cs *APIServerControllerSet) WithRevisionController(
 
 func (cs *APIServerControllerSet) WithoutRevisionController() *APIServerControllerSet {
 	cs.revisionController.controller = nil
-	cs.workloadController.emptyAllowed = true
+	cs.revisionController.emptyAllowed = true
+	return cs
+}
+
+func (cs *APIServerControllerSet) WithPruneController(
```
Comment (p0lyn0mial) [inline pkg/operator/apiserver/controllerset/apiservercontrollerset.go:284]: `WithSecretPruneController` ?

File: pkg/operator/apiserver/controller/prune/prune_controller.go (hunk 2 above)
Comment (p0lyn0mial) [reply pkg/operator/apiserver/controller/prune/prune_controller.go:41]: NewRevisionSecretPruneController ?
or
NewSecretRevisionPruneController ?

File: pkg/operator/apiserver/controller/prune/prune_controller.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+	}
+
+	return factory.New().WithInformers(
+		c.podInformer.Informer(),
+		c.secretInformer.Informer(),
+	).WithSync(c.sync).ToController("PruneController", eventRecorder.WithComponentSuffix("prune-controller"))
+}
+
+func (c *PruneController) sync(ctx context.Context, syncContext factory.SyncContext) error {
+	klog.V(5).Info("Syncing revision pruner")
```
Comment (p0lyn0mial) [inline pkg/operator/apiserver/controller/prune/prune_controller.go:66]: log the namespace ? (could help in troubleshooting)

File: pkg/operator/apiserver/controller/prune/prune_controller.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+	podGetter corev1client.PodsGetter,
+	informers v1helpers.KubeInformersForNamespaces,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	c := &PruneController{
+		targetNamespace: targetNamespace,
+		secretPrefixes:  secretPrefixes,
+
+		secretGetter:   secretGetter,
+		podGetter:      podGetter,
```
Comment (p0lyn0mial) [inline pkg/operator/apiserver/controller/prune/prune_controller.go:54]: not used ?

File: pkg/operator/apiserver/controller/prune/prune_controller.go [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+		podInformer:    informers.InformersFor(targetNamespace).Core().V1().Pods(),
+		secretInformer: informers.InformersFor(targetNamespace).Core().V1().Secrets(),
+	}
+
+	return factory.New().WithInformers(
+		c.podInformer.Informer(),
+		c.secretInformer.Informer(),
+	).WithSync(c.sync).ToController("PruneController", eventRecorder.WithComponentSuffix("prune-controller"))
```
Comment (p0lyn0mial) [inline pkg/operator/apiserver/controller/prune/prune_controller.go:62]: we could update the name here as well

File: pkg/operator/apiserver/controller/prune/prune_controller.go [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+func (c *PruneController) sync(ctx context.Context, syncContext factory.SyncContext) error {
+	klog.V(5).Info("Syncing revision pruner")
+
+	pods, err := c.podInformer.Lister().Pods(c.targetNamespace).List(labels.SelectorFromSet(map[string]string{"apiserver": "true"}))
```
Comment (p0lyn0mial) [inline pkg/operator/apiserver/controller/prune/prune_controller.go:68]: factor out the selector ?

File: pkg/operator/apiserver/controller/prune/prune_controller.go (hunk 3 above)
Comment (sttts) [reply pkg/operator/apiserver/controller/prune/prune_controller.go:1]: +1 for the question

Comment (stlaz) [reply pkg/operator/apiserver/controller/prune/prune_controller.go:1]: it's not, although it was only used for apiservers with deployments (no static pods) until now

File: pkg/operator/apiserver/controller/prune/prune_controller.go (hunk 2 above)
Comment (stlaz) [reply pkg/operator/apiserver/controller/prune/prune_controller.go]: I went with `NewSecretRevisionPruneController`

File: pkg/operator/apiserver/controller/prune/prune_controller.go (hunk 3 above)
Comment (stlaz) [reply pkg/operator/apiserver/controller/prune/prune_controller.go]: moved to pkg/operator among other similar controllers

**Pull request merged**